hamilton/driver.py

```diff
@@ -638,20 +638,80 @@ def raw_execute(
         return out
 
     @capture_function_usage
-    def list_available_variables(self) -> List[Variable]:
+    def list_available_variables(
+        self, *, tag_filter: Dict[str, Union[Optional[str], List[str]]] = None
+    ) -> List[Variable]:
         """Returns available variables, i.e. outputs.
 
-        These variables corresond 1:1 with nodes in the DAG, and contain the following information:
-        1. name: the name of the node
-        2. tags: the tags associated with this node
-        3. type: The type of data this node returns
-        4. is_external_input: Whether this node represents an external input (required from outside),
-        or not (has a function specifying its behavior).
+        These variables correspond 1:1 with nodes in the DAG, and contain the following information:
+
+            1. name: the name of the node
+            2. tags: the tags associated with this node
+            3. type: The type of data this node returns
+            4. is_external_input: Whether this node represents an external input (required from outside),
+            or not (has a function specifying its behavior).
 
+        .. code-block:: python
 
+            # gets all
+            dr.list_available_variables()
+            # gets exact matching
+            dr.list_available_variables({"TAG_NAME": "TAG_VALUE"})
+            # gets exact matching -- all tags must match, order invariant
+            dr.list_available_variables({"TAG_NAME": ["TAG_VALUE1", "TAG_VALUE2"]})
+            # gets all with matching tag, irrespective of value
+            dr.list_available_variables({"TAG_NAME": None})
+
+        :param tag_filter: A dictionary of tags to filter by. Only nodes matching the tags and their values will
+            be returned. If the value for a tag is None, then we will return all nodes with that tag. If the value
+            is non-empty we will return all nodes with that tag and that value.
         :return: list of available variables (i.e. outputs).
         """
-        return [Variable.from_node(n) for n in self.graph.get_nodes()]
+
+        def exclude_node(n: node.Node, query_dict: Dict[str, Optional[str]]) -> bool:
```

Comment:
This should probably be a separate, unit-testable function?

Reply:
done.